src/main/java/org/opensearch/neuralsearch/mappingtransformer/SemanticMappingTransformer.java

```diff
@@ -0,0 +1,332 @@
+     */
+
+    @Override
+    public void transform(Map<String, Object> mapping, TransformContext context, ActionListener<Void> listener) {
+        try {
+            final Map<String, Object> properties = getProperties(mapping);
+            // If there is no property or its format is not valid we simply do nothing and rely on core to validate the
+            // mappings and handle the error.
+            if (properties == null) {
+                listener.onResponse(null);
+                return;
+            }
+
+            Map<String, Map<String, Object>> semanticFieldPathToConfigMap = new HashMap<>();
+            String rootPath = Strings.EMPTY;
+            SemanticMappingUtils.collectSemanticField(properties, rootPath, semanticFieldPathToConfigMap);
+
+            fetchModelAndModifyMapping(semanticFieldPathToConfigMap, properties, listener);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> getProperties(Map<String, Object> mapping) {
+        if (mapping == null) {
+            return null;
+        }
+        // Actions like create index and legacy create/update index template will have the mapping properties under a
+        // _doc key. Other actions like update mapping and create/update index template will not have the _doc layer.
+        if (mapping.containsKey(DOC) && mapping.get(DOC) instanceof Map) {
+            Map<String, Object> doc = (Map<String, Object>) mapping.get(DOC);
+            if (doc.containsKey(PROPERTIES) && doc.get(PROPERTIES) instanceof Map) {
+                return (Map<String, Object>) doc.get(PROPERTIES);
+            } else {
+                return null;
+            }
+        } else if (mapping.containsKey(PROPERTIES) && mapping.get(PROPERTIES) instanceof Map) {
+            return (Map<String, Object>) mapping.get(PROPERTIES);
+        } else {
+            return null;
+        }
+    }
+
+    private void fetchModelAndModifyMapping(
+        @NonNull final Map<String, Map<String, Object>> semanticFieldPathToConfigMap,
+        @NonNull final Map<String, Object> mappings,
+        @NonNull final ActionListener<Void> listener
+    ) {
+        final Map<String, List<String>> modelIdToFieldPathMap = SemanticMappingUtils.extractModelIdToFieldPathMap(
+            semanticFieldPathToConfigMap
+        );
+        if (modelIdToFieldPathMap.isEmpty()) {
+            listener.onResponse(null);
+        }
+        final AtomicInteger counter = new AtomicInteger(modelIdToFieldPathMap.size());
+
+        for (String modelId : modelIdToFieldPathMap.keySet()) {
+            mlClientAccessor.getModel(modelId, new ActionListener<>() {
+                @Override
+                public void onResponse(MLModel mlModel) {
+                    try {
+                        modifyMappings(mappings, mlModel, modelIdToFieldPathMap.get(modelId), semanticFieldPathToConfigMap, modelId);
```

Comment:
Any multithreading concerns here with modifying a map like this if the map isn't a ConcurrentHashMap/SkipListMap? I An alternative thread-safe way could be to use PlainActionListeners to await the model info with blocking and then modify the mappings synchronously